src/ApiCheck/ApiListingGenerator.cs

```diff
@@ -0,0 +1,409 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using ApiCheck.Description;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace ApiCheck
+{
+    public class ApiListingGenerator
+    {
+        private const BindingFlags SearchFlags = BindingFlags.Public |
+            BindingFlags.NonPublic |
+            BindingFlags.Instance |
+            BindingFlags.Static |
+            BindingFlags.DeclaredOnly;
+
+        private readonly Assembly _assembly;
+        private readonly IEnumerable<Func<MemberInfo, bool>> _filters;
+
+        public ApiListingGenerator(Assembly assembly, IEnumerable<Func<MemberInfo, bool>> filters)
+        {
+            _assembly = assembly;
+            _filters = filters;
+        }
+
+        public static JObject GenerateApiListingReport(Assembly assembly, IEnumerable<Func<MemberInfo, bool>> filters = null)
+        {
+            var generator = new ApiListingGenerator(assembly, filters ?? Enumerable.Empty<Func<MemberInfo, bool>>());
+            var ApiListingDocument = generator.GenerateApiListing();
+            return JObject.FromObject(ApiListingDocument);
+        }
+
+        public ApiListing GenerateApiListing()
+        {
+            var types = _assembly.DefinedTypes;
+
+            var document = new ApiListing();
+            document.AssemblyIdentity = _assembly.GetName().ToString();
+
+            foreach (var type in _assembly.DefinedTypes.Where(type => _filters.All(filter => !filter(type))))
+            {
+                var ApiListingType = GeneratetypeDescriptor(type);
+                document.Types.Add(ApiListingType);
+            }
+
+            return document;
+        }
+
+        private TypeDescriptor GeneratetypeDescriptor(TypeInfo type)
+        {
+            var typeDescriptor = new TypeDescriptor();
+
+            typeDescriptor.Name = TypeDescriptor.GetTypeNameFor(type);
+
+            typeDescriptor.Kind = type.IsInterface ? TypeKind.Interface :
+                !type.IsValueType ? TypeKind.Class :
+                type.IsEnum ? TypeKind.Enumeration :
+                TypeKind.Struct;
```

Comment:
I think the right answer for this is to have tests (which we have), not an extra value. That said, it's cheap so I've added an extra Undefined value and throw immediately after.